hoomd/BoxDim.h

```diff
@@ -496,6 +669,128 @@ struct
+        Scalar3 origin = (m_hi + m_lo) / Scalar(2.0);
+
+        // tilt factors for nonperiodic boxes are always calculated w.r.t. to the global box with
+        // origin (0,0,0)
+        if (!m_periodic.y)
+            {
+            origin.y = Scalar(0.0);
+            }
+        if (!m_periodic.z)
+            {
+            origin.z = Scalar(0.0);
+            }
+
+        if (m_periodic.x)
+            {
+            Scalar tilt_x = (m_xz - m_xy * m_yz) * (pos.z - origin.z) + m_xy * (pos.y - origin.y);
+            if (((pos.x >= m_hi.x + tilt_x) && !flags.x) || flags.x == 1)
+                {
+                pos.x -= L.x;
+                vel.x -= m_L_rate.x;
+                img.x++;
+                }
+            else if (((pos.x < m_lo.x + tilt_x) && !flags.x) || flags.x == -1)
+                {
+                pos.x += L.x;
+                vel.x += m_L_rate.x;
+                img.x--;
+                }
+            }
+
+        if (m_periodic.y)
+            {
+            Scalar tilt_y = m_yz * (pos.z - origin.z);
+            if (((pos.y >= m_hi.y + tilt_y) && !flags.y) || flags.y == 1)
+                {
+                pos.y -= L.y;
+                pos.x -= L.y * m_xy;
+                vel.y -= m_L_rate.y;
+                vel.x -= L.y * m_xy_rate;
+                img.y++;
+                }
+            else if (((pos.y < m_lo.y + tilt_y) && !flags.y) || flags.y == -1)
+                {
+                pos.y += L.y;
+                pos.x += L.y * m_xy;
+                vel.y += m_L_rate.y;
+                vel.x += L.y * m_xy_rate;
+                img.y--;
+                }
+            }
+
+        if (m_periodic.z)
+            {
+            if (((pos.z >= m_hi.z) && !flags.z) || flags.z == 1)
+                {
+                pos.z -= L.z;
+                pos.y -= L.z * m_yz;
+                pos.x -= L.z * m_xz;
+                vel.z -= m_L_rate.z;
+                vel.y -= L.z * m_yz_rate;
+                vel.x -= L.z * m_xz_rate;
+                img.z++;
+                }
+            else if (((pos.z < m_lo.z) && !flags.z) || flags.z == -1)
+                {
+                pos.z += L.z;
+                pos.y += L.z * m_yz;
+                pos.x += L.z * m_xz;
+                vel.z += m_L_rate.z;
+                vel.y += L.z * m_yz_rate;
+                vel.x += L.z * m_xz_rate;
+                img.z--;
+                }
+            }
+        }
+
+    //! Wrap a vec3
+    HOSTDEVICE void
+    wrap(vec3<Scalar>& pos, vec3<Scalar>& vel, int3& img, char3 flags = make_char3(0, 0, 0)) const
+        {
+        Scalar3 pos_scalar = vec_to_scalar3(pos);
+        Scalar3 vel_scalar = vec_to_scalar3(vel);
+        wrap(pos_scalar, vel_scalar, img, flags);
+        pos.x = pos_scalar.x;
+        pos.y = pos_scalar.y;
+        pos.z = pos_scalar.z;
+        vel.x = vel_scalar.x;
+        vel.y = vel_scalar.y;
+        vel.z = vel_scalar.z;
```

Comment:
Can you do the same here as in `minImage`?
```suggestion
        pos = vec3<Scalar>(pos_scalar);
        vel = vec3<Scalar>(vel_scalar);
```